### PR TITLE
[ENH] Add topographic map feature via MNE

### PR DESCRIPTION
**Pull request by anandsaini024**
Script to generate the topomap feature.

### CODE REVIEW

File: mne_icalabel/features/topomap.py
Comment (mscheltienne) [inline mne_icalabel/features/topomap.py:23]:
```suggestion
```

File: mne_icalabel/features/topomap.py
Comment (mscheltienne) [inline mne_icalabel/features/topomap.py:34]:
```suggestion
```

**Comment by mscheltienne** [comment]
Hey @anandsaini024, for starters, try to remove a bit all the extra-blank lines everywhere; 

and for the test(s), start with a very simple one that just loads a dataset (from the mne testing dataset), runs an ICA decomposition, and runs this function. And you can just start with an assert on the type and shape returned. Then we will think about more advanced tests to ensure it works correctly.

File: mne_icalabel/features/topomap.py
Comment (mscheltienne) [inline mne_icalabel/features/topomap.py:68]:
```suggestion
    """Generate a topographic map as image.
```

File: mne_icalabel/features/topomap.py
Comment (mscheltienne) [inline mne_icalabel/features/topomap.py:77]:
```suggestion
    vmax : float | callable | None
```

File: mne_icalabel/features/topomap.py
Comment (mscheltienne) [inline mne_icalabel/features/topomap.py:80]:
```suggestion
    res : int
        The resolution of the square topomap image (in pixels).
```

File: mne_icalabel/features/topomap.py
Comment (mscheltienne) [inline mne_icalabel/features/topomap.py:82]:
```suggestion
    image_interp : str
```

File: mne_icalabel/features/topomap.py
Comment (mscheltienne) [inline mne_icalabel/features/topomap.py:89]:
```suggestion
    Returns
    -------
```

File: mne_icalabel/features/topomap.py
Comment (mscheltienne) [inline mne_icalabel/features/topomap.py:91]:
```suggestion
    Zi : array of size (n_pixels, n_pixels)
        Topographic map array
```

**Review by mscheltienne — CHANGES_REQUESTED** [review]
Let's start with those items, and please add the simple test to just run the function on a testing dataset file. At least this way all the codecov warning will be gone and it will be easier to review the functions.

File: mne_icalabel/features/topomap.py
Comment (anandsaini024) [reply mne_icalabel/features/topomap.py:62]: I will add this in further modification after simple tests.

File: mne_icalabel/features/tests/test_features.py
Comment (mscheltienne) [inline mne_icalabel/features/tests/test_features.py:21]: Some more conventions, 2 lines between functions ;) If you run from the root of the repository `black mne_icalabel` it will autoformat this kind of style convention.
```suggestion
from mne.datasets import testing
from mne.io import read_raw
from mne.preprocessing import ICA
from numpy.typing import NDArray

from mne_icalabel.features import topomap


def _create_test_inst_and_ica():
    directory = testing.data_path() / "MEG" / "sample"
    raw = read_raw(directory / "sample_audvis_trunc_raw.fif", preload=False)
    raw.pick_types(eeg=True)
    raw.load_data()
    ica = ICA(n_components=5, method="picard")
    ica.fit(raw)
    return ica, raw


def test_topomap_feature():
    """Test scalp topography array generation"""
    ica, inst = _create_test_inst_and_ica()
    topo_array = topomap.get_topo_array(ica, picks="eeg")
    assert isinstance(topo_array, NDArray)
    assert topo_array.shape == (ica.n_components_, 64, 64)
```

File: mne_icalabel/features/tests/test_features.py
Comment (mscheltienne) [inline mne_icalabel/features/tests/test_features.py:20]: That would actually work, but type hints are designed to be used in the declaration of a method/funcrtion/class for the arguments and the returns. 
If you want to check that a variable is of a given type, you should compare directly against it. In this case: `isinstance(topo_array, np.ndarray)`.

File: mne_icalabel/features/tests/test_features.py
Comment (mscheltienne) [inline mne_icalabel/features/tests/test_features.py:14]: Nit-picking again: this approach you choose is very good if you have a code snippet that is repeated in multiple tests.
For now, there is only one test in this file, so there is no need to define this 'loading' step in a separate function.

Also, let's suppose we have multiple tests, e.g. 5, instead of calling 5 times `_create_test_inst_and_ica`, it would be better to load only once and fit only once an ICA, because this step is time-consuming and will impact the speed at which tests runs. Thus, usually across the MNE tests, the dataset and any needed fit is done only once, at the same indentation level as the imports at the top of the test file.

Example: https://github.com/mne-tools/mne-python/blob/6955f977ac12612f38ab30277dcfff94315a0463/mne/preprocessing/tests/test_ica.py#L36-L56

All those variables are defined in the module (file) scope and are accessible directly within all the functions defined in that same module (file).

**Comment by anandsaini024** [comment]
This happened when I tried to rename the branch.

**Comment by mscheltienne** [comment]
@anandsaini024 No need to rename the branch, I was asking to rename the test file from `test_features.py` to `test_topomap.py`.

File: mne_icalabel/features/tests/test_features.py
Comment (anandsaini024) [reply mne_icalabel/features/tests/test_features.py:14]: Okay, I thought we are going to write one script to test all the features. Will rename it.

File: mne_icalabel/features/__init__.py
Comment (mscheltienne) [inline mne_icalabel/features/__init__.py:2]:
```suggestion
from .topomap import topographic_map
```

Here you need a relative import. You are in the `features` modules (which corresponds to `.`) and you need to import `topographic_map` from `topomap.py`. So `.topomap` -> `topomap.py` file in the current module (`features`).
That will also fix the tests and run them.

File: mne_icalabel/features/topomap.py
Comment (mscheltienne) [inline mne_icalabel/features/topomap.py:29]:
```suggestion
def get_topomaps(ica: ICA, picks=None):
    """Generate an array of scalp topographies (n_pixels, n_pixels) for the picked components.

    Parameters
    ----------
    ica : ICA
        Instance of MNE `~mne.preprocesssing.ICA` decomposition.
    %(picks_ica)s ``None`` (default) will pick all sources in the order fitted.

    Returns
    -------
    topomaps : array of shape (n_components, n_pixels, n_pixels)
    """
```

Let's try to be consistent with the names to make the API simpler:
`topomap.py` so let's consider this feature's name as 'topomap'. Let's go for:

- `get_topomap` for single IC -> (n_pixels, n_pixels)
- `get_topomaps` for multiple ICs at once -> (n_components, n_pixels, n_pixels)

Also, the `picks` here should refer to the `picks_ica` to pick ICs: https://github.com/mne-tools/mne-python/blob/e8ded9719de7425c8d6752d2942a6ecba20fd0fc/mne/utils/docs.py#L2289-L2292
So the default to `eeg` and the docstring is wrong as it is not used to pick data channels but ICs. I think it's also missing the logic below to use the picks. Those lines should be adapted:

```
    data = np.dot(
        ica.mixing_matrix_[:, : ica.n_components_].T,
        ica.pca_components_[: ica.n_components_],
    )
```

One more note, in the docstring I changed the ICA parameter line to 
```
Instance of MNE `~mne.preprocesssing.ICA` decomposition.
``` 

which will render as `Instance of MNE ICA decomposition.` with a x-link to the ICA API documentation. The x-link is handled by the backticks, and the tilde `~` removes the parent modules (mne, preprocessing) and only keeps the class/function/method name.

File: mne_icalabel/features/topomap.py
Comment (mscheltienne) [inline mne_icalabel/features/topomap.py:43]:
```suggestion
```

File: mne_icalabel/features/topomap.py
Comment (mscheltienne) [inline mne_icalabel/features/topomap.py:46]:
```suggestion
```

File: mne_icalabel/features/topomap.py
Comment (mscheltienne) [inline mne_icalabel/features/topomap.py:45]:
```suggestion
        np.nan_to_num(topo_, nan=0.0, copy=False)
```

For code readability, explicitly write what nan is converted to.

File: mne_icalabel/features/topomap.py
Comment (mscheltienne) [inline mne_icalabel/features/topomap.py:53]:
```suggestion
def get_topomap(
```

For API consistency, as explained above.

File: mne_icalabel/features/topomap.py
Comment (mscheltienne) [inline mne_icalabel/features/topomap.py:54]:
```suggestion
    data: NDArray[float],
```

File: mne_icalabel/features/topomap.py
Comment (mscheltienne) [inline mne_icalabel/features/topomap.py:62]:
```suggestion
    image_interp=_INTERPOLATION_DEFAULT,  # 'cubic'
    border=_BORDER_DEFAULT,  # 'mean'
    extrapolate=_EXTRAPOLATE_DEFAULT,  # 'head'
```

Nit-picking, by convention the comments are written with 2 spaces before the `#` (as you did) and one after.
e.g. `code  # comment`.

File: mne_icalabel/features/topomap.py
Comment (mscheltienne) [inline mne_icalabel/features/topomap.py:64]:
```suggestion
    """Generate a scalp topographic map (n_pixels, n_pixels).
```

I'm mostly removing the words 'image' and 'plot' since they relate to visuals while we are just interested in the underlying data array.

File: mne_icalabel/features/topomap.py
Comment (mscheltienne) [inline mne_icalabel/features/topomap.py:70]:
```suggestion
    pos : `mne.Info`
        Instance of `mne.Info` with the montage associated with the (n_channels,) points.
```

File: mne_icalabel/features/topomap.py
Comment (mscheltienne) [inline mne_icalabel/features/topomap.py:76]:
```suggestion
    res : int = 64
        The resolution of the square topographic map (in pixels).
```

File: mne_icalabel/features/topomap.py
Comment (mscheltienne) [inline mne_icalabel/features/topomap.py:69]:
```suggestion
    data : array of shape (n_channels,)
        The data points used to generate the topographic map.
```

File: mne_icalabel/features/topomap.py
Comment (mscheltienne) [inline mne_icalabel/features/topomap.py:80]: I think I would just rename this argument to interpolation, or interp, or interp_method.

File: mne_icalabel/features/topomap.py
Comment (mscheltienne) [inline mne_icalabel/features/topomap.py:86]:
```suggestion
    topomap : array of size (n_pixels, n_pixels)
```

Please update the variable name in the code accordingly.

File: mne_icalabel/features/topomap.py
Comment (mscheltienne) [reply mne_icalabel/features/topomap.py:29]: I also renamed the returned variable to `topomaps`, please update the variable name in the code accordingly.

File: mne_icalabel/features/topomap.py
Comment (mscheltienne) [inline mne_icalabel/features/topomap.py:96]:
```suggestion
        raise ValueError("The number of channels in the Info object and in the data array do not match.")
```

File: mne_icalabel/features/topomap.py
Comment (mscheltienne) [inline mne_icalabel/features/topomap.py:103]:
```suggestion
    # inferring (x, y) coordinates form mne.Info instance
```

File: mne_icalabel/features/topomap.py
Comment (mscheltienne) [inline mne_icalabel/features/topomap.py:111]: This check looks like a duplicate of the one above (ValueError: The number of channels ...), or am I missing something?

File: mne_icalabel/features/topomap.py
Comment (mscheltienne) [inline mne_icalabel/features/topomap.py:121]:
```suggestion
    return interp.set_locations(Xi, Yi)()  # Zi, topomap of shape (n_pixels, n_pixels)
```

Simpler.

**Review by mscheltienne — CHANGES_REQUESTED** [review]
Please have a look at the comments, it's almost done ;)
The biggest point to address is the `picks` argument in the `get_topo_array` (renamed to `get_topomaps`) function which does not relate to data channels but to ICs. Currently, neither the docstring or code are correct in that regard.

File: mne_icalabel/features/__init__.py
Comment (mscheltienne) [reply mne_icalabel/features/__init__.py:2]: With the new names, it would be 

```suggestion
from .topomap import get_topomap, get_topomaps  # noqa: F401
```

I'm also adding the `noqa` flag to disable the F401 error raised by static code analysis tool. c.f. https://www.flake8rules.com/rules/F401.html

File: mne_icalabel/features/tests/__init__.py
Comment (mscheltienne) [inline mne_icalabel/features/tests/__init__.py:1]:
```suggestion
```

File: mne_icalabel/features/tests/test_topomap_feature.py
Comment (mscheltienne) [inline mne_icalabel/features/tests/test_topomap_feature.py:6]:
```suggestion
from mne_icalabel.features import get_topomap, get_topomaps
```

File: mne_icalabel/features/tests/test_topomap_feature.py
Comment (mscheltienne) [inline mne_icalabel/features/tests/test_topomap_feature.py:1]:
```suggestion
import numpy as np
from mne.datasets import testing
```

File: mne_icalabel/features/tests/test_topomap_feature.py
Comment (mscheltienne) [inline mne_icalabel/features/tests/test_topomap_feature.py:20]:
```suggestion
def test_topomap_defaults():
    """Test scalp topography array generation"""
    topomaps = get_topomaps(ica, picks="eeg")
    assert isinstance(topomaps, np.ndarray)
    assert topomaps.shape == (ica.n_components_, 64, 64)
    
    # test single topo with fake data
    data = np.random.randint(1, 10, len(raw.ch_names))
    topomap = get_topomap(data, raw.info)
    assert isinstance(topomap, np.ndarray)
    assert topomap.shape == (64, 64)
```

File: mne_icalabel/features/tests/test_topomap_feature.py
Comment (mscheltienne) [inline mne_icalabel/features/tests/test_topomap_feature.py:4]:
```suggestion
```

File: mne_icalabel/features/topomap.py
Comment (mscheltienne) [inline mne_icalabel/features/topomap.py:86]:
```suggestion
        Topographic map array.
```

**Review by mscheltienne — CHANGES_REQUESTED** [review]

**Comment by mscheltienne** [comment]
@anandsaini024 I pushed a couple more style changes. It looks like the error in the test is related to the `picks`.

**Comment by mscheltienne** [comment]
Depends on #75 to fix the `requires_version` decorator and to add the `fill_doc` and `copy_doc` decorators.

File: mne_icalabel/features/topomap.py
Comment (adam2392) [inline mne_icalabel/features/topomap.py:141]: Should this be in utils, and maybe add an argument to make it generalize for any version?

File: mne_icalabel/features/topomap.py
Comment (adam2392) [inline mne_icalabel/features/topomap.py:28]: It seems like `get_topomaps` works on ICA instance, whereas `get_topomap` works on a np.ndarray. Perhaps in line with the MNE tfr module naming, we can make it more explicit like:

`get_topomaps` stays the same
`get_topomap` renamed to something like `get_topomap_array` to make it explicit that it works on an array?

Not wedded to this idea, so just a thought.

**Review by adam2392 — COMMENTED** [review]
I left a few comments, but overall this LGTM once we address those.

File: mne_icalabel/features/topomap.py
Comment (mscheltienne) [reply mne_icalabel/features/topomap.py:141]: Usually, I would say yes, but this code should be removed once we bump the MNE requirement from 1.0 to 1.1. 
I think keeping this 'temporary' code in the same file makes this future step easier, even if in this case I don't foresee this req. bump to happen anytime soon.

I don't think we need a general case for this function, we won't use it anyway and we should support all the versions of MNE included in the requirements.

By the way, this topographic feature requires MNE 1.1 because of recent changes and bug fixes in the interpolation. c.f. https://github.com/mne-tools/mne-python/pull/10579 and https://github.com/mne-tools/mne-python/pull/10617

File: mne_icalabel/features/topomap.py
Comment (mscheltienne) [reply mne_icalabel/features/topomap.py:28]: Done, makes more sense indeed!

File: mne_icalabel/features/topomap.py
Comment (adam2392) [reply mne_icalabel/features/topomap.py:141]: Okay fair. I'm in favor of just automatically going to only support v1.1+ for mne-icalabel v0.3 since we're in super-fast dev mode.

Maybe add a `#TODO` comment then to remove this when v1.1 is released?

File: mne_icalabel/features/topomap.py
Comment (adam2392) [inline mne_icalabel/features/topomap.py:138]:
```suggestion
# TODO: remove this and the corresponding import when mne v1.1 is released
def _check_mne_version():
```

**Review by adam2392 — APPROVED** [review]

File: mne_icalabel/features/topomap.py
Comment (mscheltienne) [reply mne_icalabel/features/topomap.py:142]: That works for me!

**Review by mscheltienne — APPROVED** [review]
LGTM after a couple more small changes ;) I'll merge if CIs are green.

**Comment by mscheltienne** [comment]
Thanks @anandsaini024 ! All done for this one, you can now update the branch from the PSD feature PR.

**Pull request merged**
